### PR TITLE
[KO-344] Montoring stack

### PR DESCRIPTION
**Pull request by tanmayja**
A consolidated example to setup full monitoring stack(grafana dashboards, prometheus and alert manager).

### CODE REVIEW

File: config/monitoring/alertmanager/service.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    app: aerospike-monitoring-stack-alertmanager
+    chart: aerospike-monitoring-stack
+    release: aerospike-monitoring-stack
+    app.kubernetes.io/component: aerospike-alertmanager
+spec:
+  ports:
+    - name: http
+      port: 9093
+  selector:
+    app: aerospike-monitoring-stack-alertmanager
```
Comment (abhishekdwivedi3060) [inline config/monitoring/alertmanager/service.yaml:16]: This selector doesn't match with STS pods

File: config/monitoring/alertmanager/kustomization.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+labels:
+  - includeSelectors: true
+    pairs:
+      app.kubernetes.io/component: aerospike-alertmanager
```
Comment (abhishekdwivedi3060) [inline config/monitoring/alertmanager/kustomization.yaml:7]: If we are defining this section, we don't have to add this label in all the resource yamls individually. 
It'll be easy to change from one place. 
Do this change for all the components wherever applicable

File: config/monitoring/prometheus/statefulset.yaml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+          livenessProbe:
+            httpGet:
+              path: /-/healthy
+              port: 9090
+            initialDelaySeconds: 30
+            periodSeconds: 10
+            timeoutSeconds: 10
+            successThreshold: 1
+            failureThreshold: 3
+          resources:
```
Comment (sud82) [inline config/monitoring/prometheus/statefulset.yaml:43]: No need for empty object

File: config/monitoring/prometheus/statefulset.yaml [hunk 4]
```diff
@@ -0,0 +1,76 @@
+apiVersion: apps/v1
```
Comment (sud82) [inline config/monitoring/prometheus/statefulset.yaml:1]: Check comments from the previous STSs

**Review by sud82 — CHANGES_REQUESTED** [review]

File: config/monitoring/alertmanager/statefulset.yaml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+kind: StatefulSet
+metadata:
+  name: aerospike-monitoring-stack-alertmanager
+  namespace: monitoring
+  labels:
+    app: aerospike-monitoring-stack-alertmanager
+    chart: aerospike-monitoring-stack
+    release: aerospike-monitoring-stack
+    app.kubernetes.io/component: aerospike-alertmanager
+    unique-app: aerospike-monitoring-stack-alertmanager
```
[the remark this answers is not in the document]
Comment (tanmayja) [reply config/monitoring/alertmanager/statefulset.yaml]: removed already.

File: config/monitoring/grafana/statefulset.yaml [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+          image: "grafana/grafana:latest"
+          imagePullPolicy: "IfNotPresent"
+          volumeMounts:
+            - name: grafana-config
+              mountPath: "/etc/grafana/"
+            - name: grafana-provisioning-datasources
+              mountPath: "/etc/grafana/provisioning/datasources"
+            - name: grafana-dashboards
+              mountPath: "/etc/grafana/provisioning/dashboards"
+            - name: cluster-dashboard
```
[the remark this answers is not in the document]
Comment (tanmayja) [reply config/monitoring/grafana/statefulset.yaml:24]: when created single configmap, faces the max size issue. Hence created separate mounts.

File: config/monitoring/alertmanager/statefulset.yaml [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+            httpGet:
+              path: /-/ready
+              port: 9093
+          volumeMounts:
+            - mountPath: /etc/alertmanager
+              name: alertmanagerconf
+            - mountPath: /alertmanager
+              name: alertmanagerdata
+      securityContext:
+        fsGroup: 26
```
[the remark this answers is not in the document]
Comment (tanmayja) [reply config/monitoring/alertmanager/statefulset.yaml:35]: runAsNonRoot is not allowed in any of 3 components.
alertmanager error:
`Error: container has runAsNonRoot and image has non-numeric user (nobody), cannot verify user is non-root (pod: "aerospike-monitoring-stack-alertmanager-0_monitoring(4b639449-8ae7-4fd8-9aa3-83abe7b2d41d)", container: alertmanager)`

File: config/monitoring/alertmanager/statefulset.yaml [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+  name: aerospike-monitoring-stack-alertmanager
+  namespace: monitoring
+  labels:
+    app: aerospike-monitoring-stack-alertmanager
+    chart: aerospike-monitoring-stack
+    release: aerospike-monitoring-stack
+    app.kubernetes.io/component: aerospike-alertmanager
+    unique-app: aerospike-monitoring-stack-alertmanager
+spec:
+  selector: {}
```
[the remark this answers is not in the document]
Comment (tanmayja) [reply config/monitoring/alertmanager/statefulset.yaml]: selectors will be populated by kustomization.
```
labels:
  - includeSelectors: true
    pairs:
      app.kubernetes.io/component: aerospike-alertmanager
      app: aerospike-monitoring-stack-alertmanager
```

**Review by abhishekdwivedi3060 — APPROVED** [review]

**Review by sud82 — APPROVED** [review]

File: config/monitoring/grafana/statefulset.yaml [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+            failureThreshold: 10
+          env:
+            - name: GF_SECURITY_ADMIN_USER
+              value: "admin"
+            - name: GF_SECURITY_ADMIN_PASSWORD
+              value: "admin"
+            - name: GF_PATHS_DATA
+              value: /data/grafana/data
+      securityContext:
+        fsGroup: 472
```
[the remark this answers is not in the document]
Comment (tanmayja) [reply config/monitoring/grafana/statefulset.yaml:69]: It is needed to set fsgroup as current user in grafana is uid=472(grafana) and we get error
Error: ✗ failed to connect to database: mkdir /data/grafana: permission denied

**Pull request merged**